### ISSUE TITLE
Ability to transfer files by dnd and copy/paste

### ISSUE THREAD

**Issue by TingPing** [flatpak/xdg-desktop-portal#99]
If you drag a file into a sandboxed application the path will be on the host where you won't have access.

**Comment by matthiasclasen** [comment]
This will probably need work in the compositor: it needs to recognize that the drop target is sandboxed, and make sure that the dropped files are accessible to it (by exporting it in the document portal, if necessary)

**Comment by matthiasclasen** [comment]
Or maybe this should be up to the drag source ?

**Comment by alexlarsson** [comment]
In X, is anything but the two interating apps involved in the operation at all?

**Comment by alexlarsson** [comment]
I guess in the wayland case the compositor is more involved?

**Comment by matthiasclasen** [comment]
In the wayland case, the compositor is the middle man for any of these interactions. In X, yes. this is  direct client-to-client communication by means of selections

**Comment by matthiasclasen** [comment]
One way to go about this would be to only enable it in wayland, and argue that X is unsuitable for sandboxing...

**Comment by alexlarsson** [comment]
We could make Gtk do the l export on the dropping side if it somehow detects that the peer is a flatpak app. However, we would then need some kind of X property on toplevels for flatpak app, and I don't really know how you would do that.

I guess we need something similar for the wayland case too. Or i guess the compositor knows the pid of the app and can tell if its sandboxed.

**Comment by matthiasclasen** [comment]
A fairly elegant way to handle would be to introduce a mime type for this. Instead of requesting text/uri-list, as we currently do, we could request text/x-document-portal-uri-list or somesuch when in a sandbox, and gtk could learn to treat this by:

1) export the files in the document store
2) put the document paths in the request
3) treat it like text/uri-list otherwise

**Comment by matthiasclasen** [comment]
Seems that I've overlooked that we do need to know the target app-id in order to export the files, and that probably has to be provided by the compositor, to be trustworthy. So a fixed mime-type won't work. We could do something like text/uri-list;appid=org.gnome.Recipes, but since we will have to rewrite the mime-type in the compositor anyway, we might as well handle the exporting in the compositor.

**Comment by alexlarsson** [comment]
Could be a very general compositor feature though, like it mapping toplevel to appid. Is there a way to guarantee you're talking to the X compositor or WM though and not some random client?

**Comment by alexlarsson** [comment]
Alternatively we could just send the entire file over the dnd protocol. Or is that  a horrible idea?
Mimetype "application/x-file-contents".

**Comment by TingPing** [comment]
I DnD 7GB movie files all the time, I cannot imagine that ending well.

**Comment by matthiasclasen** [comment]
https://gitlab.gnome.org/GNOME/gnome-shell/issues/418

**Comment by muelli** [comment]
> Alternatively we could just send the entire file over the dnd protocol.

that'd probably work for us. We have relatively small files in the range of a few kilobytes.

**Comment by matthiasclasen** [comment]
You can do that today. Just use a different mime type and send the content

**Comment by muelli** [comment]
more precisely, I want to drag an attachment from Evolution running on the host into an app running inside a flatpak container.  So I'm not controlling the sending side in this case. Is it still possible to send the file contents somehow?

**Comment by matthiasclasen** [comment]
Not without changing the sender

**Comment by matthiasclasen** [comment]
The existing file forwarding functionality of flatpak can be pressed into service for this.
Assuming the shell manages a text/uri-list copy/paste operation and is about to deliver it to
application XYZ. running

flatpak run --command=/usr/bin/echo --file-forwading XYZ @@ file1 file2 ... @@

will produce a rewrittten list of files on stdout that can be passed on to XYZ. Example:

flatpak run --command=/usr/bin/echo --file-forwarding org.gnome.Todo @@ /etc/passwd ~/test.c @@
/run/user/1000/doc/cca9e5f3/passwd /home/mclasen/test.c

This is of course pretty inefficient, and we should add a way to do this rewrite without launching a container.

**Comment by matthiasclasen** [comment]
https://github.com/matthiasclasen/flatpak/tree/document-forwarding has a command that shows exactly what a compositor needs to do to rewrite a list of files.

**Comment by matthiasclasen** [comment]
See https://github.com/flatpak/flatpak/pull/2065 for some musings about the best way forward.

**Comment by matthiasclasen** [comment]
I've merged the file-transfer portal. we will use that in the toolkit to implement file c/p and dnd

**Comment by muelli** [comment]
cool. how can an app make use of that portal?

**Comment by matthiasclasen** [comment]
The main use case is file copy/paste or dnd. Gtk4 has support for it already. We should find somebody to back port it to gtk3. And I'll look at adding api for it to libportal. In the meantime, the portal-test-app example in my github shows how to use the portal directly

**Comment by muelli** [comment]
cool. Let me [link it here](https://github.com/matthiasclasen/portal-test-app/blob/5f429def90c7d5aa1d1e730d44b0fb43dfc7386e/main.c) so that I will remember in the future.
is there a Gtk4 example, by any chance?

**Comment by matthiasclasen** [comment]
When the support is built into the tookit, it should be entirely transparent - you just select a file, drag it or copy it to the clipboard, and under the covers, GTK talks to the portal to make the right thing happen. Of course, it will only work if both sides of the transfer support this, so getting it into more tookits will be important.

**Comment by dreua** [comment]
> We should find somebody to back port it to gtk3.

Any update on this? Do I understand correctly that it is currently not easily possible to support dnd copy/paste through portals in Gtk3?

**Comment by utkb** [comment]
I have similar problem in Ubuntu 20.04 with Falatpak package, i can't send files from file manager's send menu to geary. There is a  "Attach not found" error message. I am not sure wheather this is same problem or not.

**Comment by dreua** [comment]
@utkb: Your problem could be related but it's not productive to discuss it here. Please create an issue against Geary or the Geary Flatpak if that has it's own issue tracker. Include a link to this issue just in case the developers/maintainers are not aware of it.

**Comment by mjog** [comment]
It's [not a bug in Geary](https://gitlab.gnome.org/GNOME/geary/issues/562), please don't file a bug about this there.

**Comment by dreua** [comment]
@mjog @utkb Sorry, I didn't do proper research and just assumed that it wouldn't be a known limitation in Geary. In general, users should report any issues against the packages first so that the respective developers can have a look. In this case this has already happened and @utkb's Geary problem is indeed blocked by this bug.

**Comment by David-Else** [comment]
This is a massive problem with flatpak apps. I hope it gets solved very soon. I would like to drag files into SIgnal Desktop to attach but it just doesn't work. Is there a workaround? I have flatseal installed and can change permissions. Can individual apps update to GTK4 and then they will work, or will people need a whole new OS installing with GTK4?

**Comment by dreua** [comment]
There wont be a Gtk3 port, will it?

**Comment by TingPing** [comment]
While its probably technically possible for a very motivated person to get it working in GTK3 its very unlikely to happen. The DND support in GTK4 is very different to GTK3 and it would be non-trivial to port those patches over.

**Comment by Mart-Bogdan** [comment]
That's strange that all discussion is around GTK, but not a word about Qt.

I've encountered same issue with Qt based app: telegram-desktop

Saving files with save dialog is also wired, but that's another issue, I think. And it at least kinda work.

**Comment by TingPing** [comment]
Nobody here is a developer for Qt. You'll have to discuss it with them.

The portal is documented here: https://flatpak.github.io/xdg-desktop-portal/#gdbus-org.freedesktop.portal.FileTransfer

**Comment by Mart-Bogdan** [comment]
Like, I mean do there issue exist, or something? 

As I understand Flatpak should bring unified descrop experience across distros + security.

**Comment by Mart-Bogdan** [comment]
> Adds files to a session. This method can be called multiple times on a given session. Note that only regular files **(not directories)** can be added. 

Wow. So directory Drag'n'Drop don't work with flatpak?

**Comment by Mart-Bogdan** [comment]
To confirm current status.

It works only if both client and server apps are GTK4?

Perhaps someone know some aoo that definitely work? So I could check.

**Comment by Mart-Bogdan** [comment]
Yeah, they have issue for Qt: https://bugreports.qt.io/browse/QTBUG-91357

**Comment by dreua** [comment]
BTW, the Qt issue also lists a dedicated issue for GTK3. Unfortunately there is no progress either: https://gitlab.gnome.org/GNOME/gtk/-/issues/3094

**Comment by Mart-Bogdan** [comment]
GTK's issue is closed, and commit https://gitlab.gnome.org/GNOME/gtk/-/commit/9f81db139d03a1b7d9bdac298fb6424a14d645cb is merged. 

**P.S.** OH God, I wish [ForgeFed](https://forgefed.org/) was a thing, and different code hosting sites interoperate with cross links.
